granary/meetup.py

```diff
@@ -0,0 +1,102 @@
+
+    def __init__(self, access_token):
+        self.access_token = access_token
+        pass
+
+    def create(self, obj, preview=False, include_link=source.OMIT_LINK, ignore_formatting=False):
+        return self._create(obj, preview, include_link, ignore_formatting)
+
+    def post_rsvp(self, urlname, event_id, response):
+        url = API_BASE + API_RSVPS % {
+                'urlname': urlname,
+                'event_id': event_id,
+                }
+        params = 'response=%(response)s' % {
+                'response': response,
+                }
+        logging.debug('Creating RSVP=%(rsvp)s for %(urlname)s %(event_id)s' % {
+            'rsvp': response,
+            'urlname': urlname,
+            'event_id': event_id,
+            })
+        return util.urlopen(urllib.request.Request(
+            url,
+            data=params,
+            headers={
+                'Authorization': 'Bearer %(access_token)s' % {'access_token': self.access_token},
+                }
+            ))
+
+    def _create(self, obj, preview=False, include_link=source.OMIT_LINK, ignore_formatting=False):
+        if not preview in (False, True):
+            return self.return_error('Invalid Preview parameter, must be True or False')
+        verb = obj.get('verb')
+        response = None
+        if 'rsvp-yes' == verb:
+            response = 'yes'
+        elif 'rsvp-no' == verb:
+            response = 'no'
+        elif 'rsvp-maybe' == verb or 'rsvp-interested' == verb:
+            return self.return_error('Meetup.com does not support %(verb)s' % {'verb': verb})
+        else:
+            return self.return_error('Meetup.com syndication does not support %(verb)s' % {'verb': verb})
+
+        event_url = obj.get('inReplyTo')
+        if not event_url:
+            return self.return_error('missing an in-reply-to')
+
+        parsed_url_path = urllib.parse.urlparse(event_url).path
+
+        # remove the trailing slash for a Meetup URL
+        if '/' == parsed_url_path[-1]:
+            parsed_url_path = parsed_url_path[:-1]
+
+        parsed_url_parts = parsed_url_path.split('/')
+        if not 4 == len(parsed_url_parts):
+            return self.return_error('Invalid Meetup.com event URL')
+
+        urlname = parsed_url_parts[1]
+        if not urlname:
+            return self.return_error('Invalid Meetup.com event URL')
+
+        event_id = parsed_url_parts[3]
+        if not event_id:
+            return self.return_error('Invalid Meetup.com event URL')
+
+        hostname = urllib.parse.urlparse(event_url).hostname
+        if not hostname in ['www.meetup.com', 'meetup.com']:
+            return self.return_error('Invalid Meetup.com event URL')
+
+        if preview:
+            desc = ('<span class="verb">RSVP %s</span> to <a href="%s">this event</a>.' %
+                    (verb[5:], event_url))
+            return source.creation_result(description=desc)
+
+        create_resp = {
+                'url': event_url,
```

Comment:
are there rsvp permalinks? any chance we can use those? also if we get an rsvp id back, mind putting that in the 'id' field?

Reply:
Unfortunately not, this is the response from the API:

```json
{
    "created": 1576781043000,
    "updated": 1576781043000,
    "response": "yes",
    "guests": 0,
    "event": {
        "id": "267300253",
        "name": "Tech Nottingham January 2020 - Ethics In Advertising And High Performance Teams",
        "yes_rsvp_count": 35,
        "time": 1578940200000,
        "utc_offset": 0
    },
    "group": {
        "id": 14171002,
        "urlname": "Tech-Nottingham",
        "name": "Tech Nottingham",
        "status": "active",
        "who": "Techies",
        "members": 2939,
        "join_mode": "open",
        "localized_location": "Nottingham, United Kingdom",
        "group_photo": {
            "id": 460984584,
            "highres_link": "https://secure.meetupstatic.com/photos/event/1/1/e/8/highres_460984584.jpeg",
            "photo_link": "https://secure.meetupstatic.com/photos/event/1/1/e/8/600_460984584.jpeg",
            "thumb_link": "https://secure.meetupstatic.com/photos/event/1/1/e/8/thumb_460984584.jpeg",
            "type": "event",
            "base_url": "https://secure.meetupstatic.com"
        }
    },
    "member": {
        "id": 189380737,
        "name": "Jamie Tanna",
        "bio": "Software engineer at Capital One. Lover of Chef, automation and sharing knowledge.",
        "photo": {
            "id": 288326741,
            "highres_link": "https://secure.meetupstatic.com/photos/member/6/8/7/5/highres_288326741.jpeg",
            "photo_link": "https://secure.meetupstatic.com/photos/member/6/8/7/5/member_288326741.jpeg",
            "thumb_link": "https://secure.meetupstatic.com/photos/member/6/8/7/5/thumb_288326741.jpeg",
            "type": "member",
            "base_url": "https://secure.meetupstatic.com"
        },
        "event_context": {
            "host": false
        }
    },
    "venue": {
        "id": 1365520,
        "name": "Antenna",
        "lat": 52.95594787597656,
        "lon": -1.1427489519119263,
        "repinned": true,
        "address_1": "Beck Street",
        "city": "Nottingham",
        "country": "gb",
        "localized_country_name": "United Kingdom",
        "phone": "0115 993 2350"
    }
}
```

No ID for the RSVP itself, and no permalink

Reply:
disappointing. consider making up your own fragments to make permalinks? eg see how facebook.py does this for likes. maybe less important for create than for get_activities though? I'm ok either way.

Reply:
bridgy publish may actually depend on getting the id field back. not sure.

Reply:
Permalinks done with the fragment approach :+1: